### PR TITLE
feat(provider): Add Mail.ru OAuth Service Provider and Callback snippet

### PR DESCRIPTION
**Pull request by nyedidikeke**
- [x] Documentation: Update callback snippet (fixes unexpected token error on broken callbacks documentation).
- [x] Fix: Resolve Mail.ru bug #308 (missing request parameter: access_token).
- [x] Enhancement: Add in-build support for Mail.ru OAuth Service Provider.

---

**How to use Mail.ru as custom OAuth Service Provider**

```js
// pages/api/auth/[...nextauth].js
{
  id: 'mailru',
  name: 'Mail.ru',
  type: 'oauth',
  version: '2.0',
  scope: 'userinfo',
  params: {
    grant_type: 'authorization_code'
  },
  accessTokenUrl: 'https://oauth.mail.ru/token',
  requestTokenUrl: 'https://oauth.mail.ru/token',
  authorizationUrl: 'https://oauth.mail.ru/login?response_type=code',
  profileUrl: 'https://oauth.mail.ru/userinfo',
  profile: (profile) => {
    return {
      id: profile.id,
      name: profile.name,
      email: profile.email,
      image: profile.image,
    }
  },
  clientId: process.env.MAILRU_ID,
  clientSecret: process.env.MAILRU_SECRET
}
```
**How to use Mail.ru as in-build OAuth Service Provider**

```js
// pages/api/auth/[...nextauth].js
providers: [
  Providers.MailRu({
    clientId: process.env.MAILRU_ID,
    clientSecret: process.env.MAILRU_SECRET
  }),
],
```

Define both client ID and client secret values in `.env` file;
```sh
# .env
MAILRU_ID=
MAILRU_SECRET=
```

### CODE REVIEW

**Comment by nikitalk** [comment]
Hi! I've tried to signin with this provider and I've got this message in callback function:
```
{
  error: 'invalid request',
  error_code: 2,
  error_description: 'missing request parameter: access_token'
}
```

**Comment by nyedidikeke** [comment]
> Hi! I've tried to signin with this provider and I've got this message in callback function:

It's normal you get the same error @nikitalk; the pull request has not yet been merged. Once it is, you need to update `next-auth` and it should definitely work.

**Comment by nikitalk** [comment]
No, I've added this provider in my own local copy of nextauth.

**Comment by nikitalk** [comment]
lso I tried adding custom provider as below. The result is the same.

> * Fix Mail.ru bug #308 (missing request parameter: access_token).
> 
> Documentation available at https://next-auth.js.org/configuration/providers#using-a-custom-provider for configuration of a custom OAuth service provider.
> 
> How to use Mail.ru as custom OAuth service provider;
> 
> ```js
> // pages/api/auth/[...nextauth].js
> {
>   id: 'mailru',
>   name: 'Mail.ru',
>   type: 'oauth',
>   version: '2.0',
>   scope: 'userinfo',
>   params: {
>     grant_type: 'authorization_code'
>   },
>   accessTokenUrl: 'https://oauth.mail.ru/token',
>   requestTokenUrl: 'https://oauth.mail.ru/token',
>   authorizationUrl: 'https://oauth.mail.ru/login?response_type=code',
>   profileUrl: 'https://oauth.mail.ru/userinfo',
>   profile: (profile) => {
>     return {
>       id: profile.id,
>       name: profile.name,
>       email: profile.email,
>       image: profile.image,
>     }
>   },
>   setGetAccessTokenProfileUrl: true, // Should be add for Mail.ru as custom OAuth service provider.
>   clientId: process.env.MAILRU_ID,
>   clientSecret: process.env.MAILRU_SECRET
> }
> ```
> 
> Define both client ID and client secret values in `.env` file;
> 
> ```shell
> # .env
> MAILRU_ID=
> MAILRU_SECRET=
> ```

**Comment by nikitalk** [comment]
Ok. Thank you!

**Comment by nyedidikeke** [comment]
> No, I've added this provider in my own local copy of nextauth.

Take a look at [src/server/lib/oauth/callback.js](https://github.com/nextauthjs/next-auth/pull/522/commits/a4473889d3551ff69c7b15438ad451969699044d);
Making use of `setGetAccessTokenProfileUrl` will only work after a merge of the [current](https://github.com/nextauthjs/next-auth/pull/522) into the main stream of `next-auth`.

**Comment by nikitalk** [comment]
Yes, I've already seen in your fork) Thanks!

**Comment by iaincollins** [comment]
Thanks for this! The approach looks good, going to check it out over the next couple of days.

**Comment by nyedidikeke** [comment]
@iaincollins : updated current pull request as suggested in (now closed) #533 pull request review.

**Comment by nikitalk** [comment]
Hi! For correct work of this solution it is need to fix `src/server/lib/oauth/callback.js` file. In order to add access_token to `url`, it should be a variable, but not a constant:
```js
function _get (provider, accessToken, callback) {
- const url = provider.profileUrl
+ let url = provider.profileUrl
```

**Comment by nyedidikeke** [comment]
Why a constant? Ref. [const](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/const) vs [let](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/let) statements.
Even though not introduced by the current [pull](https://github.com/nextauthjs/next-auth/pull/522/commits/a4473889d3551ff69c7b15438ad451969699044d) request, it works great @nikitalk.

**Comment by nyedidikeke** [comment]
> Hi! For correct work of this solution it is needed to fix src/server/lib/oauth/callback.js file. In order to add access_token to url, it should be a variable, but not a constant:

Why a constant? Ref. [const](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/const) vs [let](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/let) statements.
Even though not introduced by the current [pull](https://github.com/nextauthjs/next-auth/pull/522/commits/a4473889d3551ff69c7b15438ad451969699044d) request, it works great @nikitalk.

**Comment by nikitalk** [comment]
> > Hi! For correct work of this solution it is needed to fix src/server/lib/oauth/callback.js file. In order to add access_token to url, it should be a variable, but not a constant:
> 
> Why a constant? Ref. [const](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/const) vs [let](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/let) statements.
> Even though not introduced by the current [pull](https://github.com/nextauthjs/next-auth/pull/522/commits/a4473889d3551ff69c7b15438ad451969699044d) request, it works great @nikitalk.

If `url` is assigned as `const`, there is an error:
image

**Comment by nikitalk** [comment]
Since `url` is `const`, it is impossible to reassign it here:
```js
if (setGetAccessTokenProfileUrl) {
      if (provider.profileUrl) {
        url = provider.profileUrl + '?access_token=' + accessToken
      }
    }
```

**Comment by nyedidikeke** [comment]
> Since `url` is `const`, it is impossible to reassign it here:
> 
> ```js
> if (setGetAccessTokenProfileUrl) {
>       if (provider.profileUrl) {
>         url = provider.profileUrl + '?access_token=' + accessToken
>       }
>     }
> ```

I get your point @nikitalk; it eluded me ... fixing that.

**Comment by nikitalk** [comment]
Yes, now is good. Thank you! I don't insist, but maybe this code will be cleaner:

```js
function _get(provider, accessToken, callback) {
  let url = provider.profileUrl;
  const headers = provider.headers || {};

  if (this._useAuthorizationHeaderForGET) {
    headers.Authorization = this.buildAuthHeader(accessToken);

    // Mail.ru requires 'access_token' as URL request parameter
    if (provider.setGetAccessTokenProfileUrl)
      url += "?access_token=" + accessToken;

    // This line is required for Twitch
    headers["Client-ID"] = provider.clientId;
    accessToken = null;
  }

  this._request("GET", url, headers, null, accessToken, callback);
}
 
```

**Comment by iaincollins** [comment]
@nyedidikeke @nikitalk Thank you both! :-)

**Review by iaincollins — CHANGES_REQUESTED** [review]
This looks great, thank you! The work on the documentation pages is much appreciated.

No issues with the pull request, just putting merging on hold briefly while we sort out the release process (so we don't merge things live on the website before we have actually released them).

Working hard to get this unblocked so we can get this rolled out!

File: src/server/lib/oauth/callback.js
Comment (whiletrueltd) [inline src/server/lib/oauth/callback.js:266]: I am afraid append url like this could be dangerous. What if the origin profileUrl contains '?' already?

Suggest to use Node.js url.searchParams.append to handle it.

Comment (wuminqi) [inline src/server/lib/oauth/callback.js:266]: I am afraid append url like this could be dangerous.  What if the origin profileUrl contains '?'  already?

Suggest to use Node.js url API searchParams.append to handle this.

Comment (balazsorban44) [reply src/server/lib/oauth/callback.js:266]: Agree.
[URLSearchParams](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams) or [URL](https://developer.mozilla.org/en-US/docs/Web/API/URL_API) would be more ideal.

Comment (ndom91) [reply src/server/lib/oauth/callback.js:266]: @nyedidikeke I think this last suggestion is also good to get in, but after that and fixing up any lingering failing tests / merge conflicts it should be good to go!

**Comment by nikitalk** [comment]
Hi, friends! Will be this solution in 3.2.0 release?

**Comment by ndom91** [comment]
@nikitalk it can be, there is still a merge conflict and a suggestion (regarding URL Params) which should be cleaned up though before we can merge.

**Comment by balazsorban44** [comment]
Just an idea💡 . Instead of increasing the user facing api by adding a new property to provider options maybe we could switch on the provider id instead, and add the access token if it's mailru.

What do you think @nyedidikeke @ndom91?

**Comment by ndom91** [comment]
Yeah good call @balazsorban44, i didn't even notice initially.

File: src/server/lib/oauth/callback.js
Comment (balazsorban44) [inline src/server/lib/oauth/callback.js:266]: instead of expanding the API surface, could you just check for `provider.id === "mailru"` instead?

**Review by balazsorban44 — CHANGES_REQUESTED** [review]
Hi there, and thanks! Please look at my comment

**Review by balazsorban44 — APPROVED** [review]
Thanks @nyedidikeke! Nice work!

**Comment by balazsorban44** [comment]
Could you please install the latest canary version and verify it works correctly, @nyedidikeke?

**Comment by nyedidikeke** [comment]
> Could you please install the latest canary version and verify it works correctly, @nyedidikeke?

It's working @balazsorban44. @nikitalk You may want to give it a try as well.

**Comment by nikitalk** [comment]
@nyedidikeke Everything is ok. Thanks a lot!!!

**Pull request merged**
